client-runtime/src/main/java/com/microsoft/rest/v2/http/AsyncInputStream.java

```diff
@@ -0,0 +1,118 @@
+    }
+
+    /**
+     * Creates an AsyncInputStream from an AsynchronousFileChannel.
+     *
+     * @param fileChannel The file channel.
+     * @param offset The offset in the file to begin reading.
+     * @param length The number of bytes of data to read from the file.
+     * @return The AsyncInputStream.
+     */
+    public static AsyncInputStream create(AsynchronousFileChannel fileChannel, long offset, long length) {
+        return new AsyncInputStream(new FlowableFileStream(fileChannel, offset, length), length);
+    }
+
+    /**
+     * Creates an AsyncInputStream from an AsynchronousFileChannel which reads the entire file.
+     * @param fileChannel The file channel.
+     * @throws IOException if an error occurs when determining file size
+     * @return The AsyncInputStream.
+     */
+    public static AsyncInputStream create(AsynchronousFileChannel fileChannel) throws IOException {
+        long size = fileChannel.size();
+        return new AsyncInputStream(new FlowableFileStream(fileChannel, 0, size), size);
+    }
+
+    /**
+     * Creates an AsyncInputStream which emits the content of a given InputStream with a known length.
+     * Will close the InputStream when subscription completes.
+     *
+     * @param inputStream The input stream.
+     * @param contentLength The length of the stream content.
+     * @return An AsyncInputStream which emits the content from the given InputStream.
+     */
+    public static AsyncInputStream create(final InputStream inputStream, long contentLength) {
+        Flowable<byte[]> content = Flowable.generate(
+                new Callable<InputStream>() {
+                    @Override
+                    public InputStream call() throws Exception {
+                        return inputStream;
+                    }
+                },
+                new BiConsumer<InputStream, Emitter<byte[]>>() {
+                    private static final int CHUNK_SIZE = 8192;
+                    private final byte[] innerBuf = new byte[CHUNK_SIZE];
+
+                    @Override
+                    public void accept(InputStream inputStream, Emitter<byte[]> emitter) throws Exception {
+                        try {
+                            int bytesRead = inputStream.read(innerBuf);
+                            if (bytesRead == -1) {
+                                emitter.onComplete();
+                            } else {
+                                byte[] nextBuf = Arrays.copyOf(innerBuf, bytesRead);
+                                emitter.onNext(nextBuf);
+                            }
+                        } catch (IOException e) {
+                            emitter.onError(e);
+                        }
+                    }
+                },
+                new Consumer<InputStream>() {
+                    @Override
+                    public void accept(InputStream inputStream) throws Exception {
+                        inputStream.close();
```

Comment:
Should we be closing this InputStream since it is provided by the user?

Reply:
Yeah, in general it seems like we should keep users in control of the lifecycle of resources they provide.